internal/nginx/modules/httpmatches.js

```diff
@@ -0,0 +1,94 @@
+export default { redirect };
+
+const matchesVariable = 'http_matches';
+
+// FIXME(osborn): Need to add special handling for repeated headers.
+// Should follow guidance from https://www.rfc-editor.org/rfc/rfc7230.html#section-3.2.2.
+function headersMatch(r, headers) {
+  for (let i = 0; i < headers.length; i++) {
+    const h = headers[i];
+    const kv = h.split(':');
+    if (kv.length !== 2) {
+      r.error('invalid header match: ' + h);
+      throw 500;
+    }
+    let val = r.headersIn[kv[0]];
+    if (!val || val.toLowerCase() !== kv[1]) {
+      r.warn('header: ' + h + ' does not exist in request');
+      throw 404;
+    }
+  }
+}
+
+function paramsMatch(r, params) {
+  for (let i = 0; i < params.length; i++) {
+    let p = params[i];
+    // get index of first = in string
+    const idx = p.indexOf('=');
+    // throw error if index is invalid
+    // an index of -1 means "=" is not present
+    // an index of 0 means there is no value
+    // an index of length -1 means there is no key
+    if (idx === -1 || (idx === 0) | (idx === p.length - 1)) {
```

Comment:
Began a discussion out of band. Valueless keys should be valid in my understanding (`?key1=&key2`). Conclusion will be remarked here.

Reply:
Out of band conclusion: API enforces a non-empty string value for "value".

Please add a quick comment to that effect since this is a few layers removed from that constraint and this isn't a general purpose parser.

Reply:
Added